### PR TITLE
OCPBUGS-5534: add polling of plugin-manifest.json to PollConsoleUpdates

### PR DESCRIPTION
**Pull request by rhamilto**
Tested by running `dynamic-demo-plugin` locally and manually editing its `plugin-manifest.json` `version` value.

Thanks to @spadgett for the guidance!

### CODE REVIEW

File: frontend/public/components/app.jsx [hunk 1]
```diff
@@ -311,41 +311,63 @@ const PollConsoleUpdates = React.memo(function PollConsoleUpdates() {
...
       .then((response) => {
         setPluginsData(response);
         setPluginsError(null);
       })
       .catch(setPluginsError);
   }, [safeFetch]);
-  usePoll(tick, URL_POLL_DEFAULT_DELAY);
+  usePoll(updatesTick, URL_POLL_DEFAULT_DELAY);
+  const fetchPluginManifest = (pluginName) =>
+    coFetchJSON(`${window.SERVER_FLAGS.basePath}api/plugins/${pluginName}/plugin-manifest.json`);
```
Comment (spadgett) [inline frontend/public/components/app.jsx:334]: I don't know if we expose this through `coFetchJSON`, but it would be good to call `fetch` with `{ cache: 'no-cache' }` to make sure we're not getting a cached copy of the manifest.

```js
  // Download a resource with cache busting when dealing with a
  // properly configured server that will send the correct ETag
  // and Date headers and properly handle If-Modified-Since and
  // If-None-Match request headers, therefore we can rely on the
  // validation to guarantee a fresh response.
  fetch("some.json", {cache: "no-cache"})
    .then(function(response) { /* consume the response */ });
```

https://hacks.mozilla.org/2016/03/referrer-and-cache-control-apis-for-fetch/

@vojtechszocs FYI

File: frontend/public/components/app.jsx [hunk 2]
```diff
@@ -311,41 +311,63 @@ const PollConsoleUpdates = React.memo(function PollConsoleUpdates() {
...
   }
 
   if (pluginsChanged && !allPluginEndpointsReady && !isFetchingPluginEndpoints) {
-    const pluginEndpointsReady = pluginsData?.plugins?.map((pluginName) => {
-      return coFetchJSON(
-        `${window.SERVER_FLAGS.basePath}api/plugins/${pluginName}/plugin-manifest.json`,
-      );
-    });
+    const pluginEndpointsReady = pluginsData?.plugins?.map((pluginName) =>
+      fetchPluginManifest(pluginName),
```
Comment (spadgett) [inline frontend/public/components/app.jsx:369]: We shouldn't need to fetch the plugin manifests again since we have them in `pluginManifestsData` / `pluginManifestsError`

File: frontend/public/components/app.jsx [hunk 3]
```diff
@@ -311,41 +311,63 @@ const PollConsoleUpdates = React.memo(function PollConsoleUpdates() {
...
+  const manifestsTick = React.useCallback(() => {
+    const pluginManifests = pluginsData?.plugins?.map((pluginName) =>
+      fetchPluginManifest(pluginName),
+    );
+    Promise.all(pluginManifests)
+      .then((response) => {
+        setPluginManifestsData(response);
+        setPluginManifestsError(null);
+      })
+      .catch(setPluginManifestsError);
```
Comment (spadgett) [inline frontend/public/components/app.jsx:344]: If any of the plugin services are down, we'll stop checking updates for any of them... It would be better to handle errors individually so that we can still use the version data we have for other plugins.

File: frontend/public/components/app.jsx [hunk 4]
```diff
@@ -358,6 +380,19 @@ const PollConsoleUpdates = React.memo(function PollConsoleUpdates() {
...
 
+  const pluginManifestsVersionsChanged = pluginManifestsData?.reduce((acc, obj) => {
+    prevPluginManifestsData?.forEach((o) => {
+      if (obj.name === o.name && obj.version !== o.version) {
+        acc.push(obj);
+      }
+    });
+    return acc;
+  }, []);
+  const pluginManifestsChanged = !_.isEmpty(pluginManifestsVersionsChanged);
```
Comment (spadgett) [inline frontend/public/components/app.jsx:391]:
```suggestion
  const pluginManifestsVersionsChanged = pluginManifestsData?.some((manifest) => {
    return prevPluginManifestsData?.some((previousManifest) => {
      return manifest.name === previousManifest.name && manifest.version !== previousManifest.version;
    });
  });
```

Comment (spadgett) [reply frontend/public/components/app.jsx:391]: This is a little inefficient with the nested loops, but since the plugin list should be small, I think it's OK.

File: frontend/public/components/app.jsx [hunk 5]
```diff
@@ -311,41 +311,63 @@ const PollConsoleUpdates = React.memo(function PollConsoleUpdates() {
...
+    prevPluginManifestsDataRef.current = pluginManifestsData;
   });
   const prevPluginsData = prevPluginsDataRef.current;
-  const stateInitialized = _.isEmpty(pluginsError) && !_.isEmpty(prevPluginsData);
+  const prevPluginManifestsData = prevPluginManifestsDataRef.current;
+  const stateInitialized =
+    _.isEmpty(pluginsError) &&
+    !_.isEmpty(prevPluginsData) &&
+    _.isEmpty(pluginManifestsError) &&
+    !_.isEmpty(prevPluginManifestsData);
```
Comment (spadgett) [inline frontend/public/components/app.jsx:360]: We should make sure we don't break the console commit version check on clusters that have no plugins. (Presumably `stateInitialized` will always be false if there are no plugins since we don't have any manifest data?)

File: frontend/public/components/app.jsx (hunk 1 above)
Comment (rhamilto) [reply frontend/public/components/app.jsx:334]: `coFetchJSON` does include an options argument, which includes `cache`.  Will update.

File: frontend/public/components/app.jsx (hunk 2 above)
Comment (spadgett) [reply frontend/public/components/app.jsx:368]: > We shouldn't need to fetch the plugin manifests again since we have them in `pluginManifestsData` / `pluginManifestsError`

Ignore this comment... After we talked, I realized this is fetching new plugins from the `/check-updates` endpoints for which we don't have the data yet.

**Review by spadgett — COMMENTED** [review]
Hey, @rhamilto. This logic is getting complicated enough that I think we need to have some tests covering it, particularly because there are some edge cases that are difficult to test manually. Maybe we can mock the HTTP responses using `cy.intercept`? Here is an example PR that does that:

https://github.com/openshift/console/pull/12421

It would be good to check that the prompt is shown when

1. A new version of console rolls out (console commit changes)
2. A new plugin is enabled, but only when the plugin manifest is there
3. A plugin is removed
4. A plugin version is updated

File: frontend/public/components/app.jsx [hunk 6]
```diff
@@ -341,11 +361,9 @@ const PollConsoleUpdates = React.memo(function PollConsoleUpdates() {
...
   if (pluginsChanged && !allPluginEndpointsReady && !isFetchingPluginEndpoints) {
-    const pluginEndpointsReady = pluginsData?.plugins?.map((pluginName) => {
-      return coFetchJSON(
-        `${window.SERVER_FLAGS.basePath}api/plugins/${pluginName}/plugin-manifest.json`,
-      );
-    });
+    const pluginEndpointsReady = pluginsData?.plugins?.map((pluginName) =>
+      fetchPluginManifest(pluginName),
+    );
     Promise.all(pluginEndpointsReady)
```
Comment (spadgett) [inline frontend/public/components/app.jsx:367]: There's an existing problem here where if one of the plugins has an error getting the manifest, we will never prompt when a new plugin is installed. I think we should track this as a separate bug, though.

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
+import { checkErrors } from '../../support';
+import { warningToast } from '../../views/form';
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts:2]: Not new, but I think `warningToast` will match any PF warning alert (inline, etc). We probably want something more specific so a different warning on the page doesn't match.

https://github.com/openshift/console/blob/1ec76caec2d8e287d03fff3c992805106aeab863/frontend/packages/integration-tests-cypress/views/form.ts#L7

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+import { guidedTour } from '../../views/guided-tour';
+
+const CHECK_UPDATES_URL = '/api/check-updates';
+const CHECK_UPDATES_ALIAS = 'checkUpdates';
+const CHECK_MANIFEST_ALIAS = 'checkManifest';
+/* The update wait is the value to wait for the poll of /api/check-updates to return with the updated list of plugins
+ after the plugin is enabled and loaded. This wait will be longer on ci than when debugging locally. */
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts:9]: I'm not sure the second sentence about taking longer in CI is accurate for these tests.

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    cy.wait(`@${CHECK_UPDATES_ALIAS}`, WAIT_OPTIONS);
+    checkConsoleUpdateToast();
+  });
+
+  it('triggers the console update toast when a plugin is added', () => {
+    loadApp();
+    cy.intercept(CHECK_UPDATES_URL, UPDATES_DEFAULT).as(CHECK_UPDATES_ALIAS);
+    cy.wait(`@${CHECK_UPDATES_ALIAS}`, WAIT_OPTIONS);
+    cy.intercept(CHECK_UPDATES_URL, UPDATES_NEW_PLUGIN).as(CHECK_UPDATES_ALIAS);
+    cy.intercept(PLUGIN_MANIFEST_URL, PLUGIN_MANIFEST_DEFAULT).as(CHECK_MANIFEST_ALIAS);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts:75]: Can we make sure the toast doesn't appear until after the plugin manifest for the new plugin is available? In fact, it would be good to return an error requesting that URL once (to simulate a plugin that has not yet been deployed).

**Review by spadgett — COMMENTED** [review]
Thank you for adding the tests 👍

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts (hunk 7 above)
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts]: Good point.  Now checking for the refresh link instead.

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts (hunk 8 above)
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts]: Good point.  Omitted the comment entirely.

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts (hunk 9 above)
Comment (rhamilto) [reply frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts:77]: Done.

File: frontend/public/components/app.jsx (hunk 6 above)
Comment (rhamilto) [reply frontend/public/components/app.jsx:367]: Opened a bug to track: https://issues.redhat.com/browse/OCPBUGS-5826

**Comment by rhamilto** [comment]
@spadgett, mind taking another look?

File: frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    cy.wait(`@${CHECK_UPDATES_ALIAS}`, WAIT_OPTIONS);
+    checkConsoleUpdateToast();
+  });
+
+  it('triggers the console update toast when a plugin is added', () => {
+    loadApp();
+    cy.intercept(CHECK_UPDATES_URL, UPDATES_DEFAULT).as(CHECK_UPDATES_ALIAS);
+    cy.wait(`@${CHECK_UPDATES_ALIAS}`, WAIT_OPTIONS);
+    cy.intercept(CHECK_UPDATES_URL, UPDATES_NEW_PLUGIN).as(CHECK_UPDATES_ALIAS);
+    cy.intercept(PLUGIN_MANIFEST_URL, { forceNetworkError: true }).as(CHECK_MANIFEST_ALIAS);
```
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/poll-console-updates.spec.ts:73]: Nice

**Review by spadgett — APPROVED** [review]
Thanks @rhamilto 👍 

/lgtm
